Enforce `opam-version` as first non-comment line for > 2.0

At present, `opam-version` is used to control the fields available in an `opam` file but it can't control the lexer. In particular, this will be limiting for adding additional operators to the constraint language.

- [x] opam-file-format / opam 2.1 need updating to reject an opam file with `opam-version: "2.1"` or greater where `opam-version` is not the first line, ignoring lines consisting of just whitespace/comments.
- [x] opam 2.0.8 should cease parsing opam files after `opam-version` is encountered (so what a 2.0 lexer would regard as syntax errors are not reported)

This is addressed in https://github.com/ocaml/opam-file-format/pull/43 and would be picked up by opam when the src_ext is updated.

Before you get errors like:

```
dra@thor:~/opam$ opam pin add opam-core .
[opam-core.2.0.8] synchronised (file:///home/dra/opam)
[ERROR] Invalid opam file in opam-core source from file:///home/dra/opam:
    error  2: File format error at line 7, column 0: '+' is not a valid token
```

now you get more informative errors like:

```
[ERROR] Invalid opam file in opam-core source from file:///home/dra/opam:
    error  2: File format error: unsupported or missing file format version; should be 2.0 or older
```

The error message for `opam-version` in the wrong place is not as nice, but this can be improved in future (I've shied away from it here because it would require changing the `OpamParser` API _again_):

```
[ERROR] Invalid opam file in opam-core source from file:///home/dra/opam:
    error  2: File format error at line 4, column 0: Parse error
```
but it does clearly point to the line with the error, which is at least no worse than any other parsing error.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -7,8 +7,8 @@ env:
   OPAMBSROOT: ~/.cache/.opam.cached
   OPAM12CACHE: ~/.cache/opam1.2/cache
   # This should be identical to the value in appveyor.yml
-  OPAM_TEST_REPO_SHA: 75ff83837547183f000e1c18d42260d7d6cc5fbc
-  OPAM_REPO_SHA: 75ff83837547183f000e1c18d42260d7d6cc5fbc
+  OPAM_TEST_REPO_SHA: b02110b549a0b5275806ce27444fabac71093a0e
+  OPAM_REPO_SHA: b02110b549a0b5275806ce27444fabac71093a0e
   # Default ocaml version for some jobs
   OCAML_VERSION: 4.11.2
   ## variables for cache paths

--- a/appveyor.yml
+++ b/appveyor.yml
@@ -12,8 +12,8 @@ environment:
     CYG_MIRROR: http://mirrors.kernel.org/sourceware/cygwin/
     DEP_MODE: lib-ext
     # This should be identical to the value in .travis.yml
-    OPAM_REPO_SHA: 75ff83837547183f000e1c18d42260d7d6cc5fbc
-    OPAM_TEST_REPO_SHA: 75ff83837547183f000e1c18d42260d7d6cc5fbc
+    OPAM_REPO_SHA: b02110b549a0b5275806ce27444fabac71093a0e
+    OPAM_TEST_REPO_SHA: b02110b549a0b5275806ce27444fabac71093a0e
   matrix:
     - CYG_ROOT: cygwin
       CYG_ARCH: x86

--- a/master_changes.md
+++ b/master_changes.md
@@ -74,6 +74,8 @@ New option/command/subcommand are prefixed with ◈.
   * Fix `features` parser [#4507 @rjbou]
   * Rename `hidden-version` to `avoid-version` [#4527 @dra27]
   * Fix rewriting with preserved format empty field error [#4634 @rjbou - fix #4628]
+  * Fix rewrtiting with preserved format empty field error [#4633 @rjbou - fix #4628]
+  * Require opam-file-format 2.1.3+ in order to enforce opam-version: "2.1" as first non-comment line [#4639 @dra27 - fix #4394]
 
 ## External dependencies
   * Handle macport variants [#4509 @rjbou - fix #4297]

--- a/opam-admin.opam
+++ b/opam-admin.opam
@@ -18,5 +18,5 @@ bug-reports: "https://github.com/ocaml/opam/issues"
 depends: [
   "dune" {>= "1.11.0"}
   "re" {>= "1.9.0"}
-  "opam-file-format" {>= "2.1.2"}
+  "opam-file-format" {>= "2.1.3"}
 ]

--- a/opam-format.opam
+++ b/opam-format.opam
@@ -27,7 +27,7 @@ build: [
 depends: [
   "ocaml" {>= "4.02.3"}
   "opam-core" {= version}
-  "opam-file-format" {>= "2.1.2"}
+  "opam-file-format" {>= "2.1.3"}
   "re" {>= "1.9.0"}
   "dune" {>= "1.11.0"}
 ]

--- a/src/format/opamFile.ml
+++ b/src/format/opamFile.ml
@@ -1016,15 +1016,24 @@ module SyntaxFile(X: SyntaxFileArg) : IO_FILE with type t := X.t = struct
   module IO = struct
     let to_opamfile filename t = Pp.print X.pp (filename, t)
 
+    let catch_future_syntax_error = function
+    | {file_contents = [{pelem = Variable({pelem = "opam-version"; _}, {pelem = String ver; _}); _ };
+                        {pelem = Section {section_kind = {pelem = "#"; _}; _}; pos}]; _}
+      when OpamVersion.(compare (nopatch (of_string ver)) (nopatch X.format_version)) <= 0 ->
+        raise (OpamPp.Bad_format (Some pos, "Parse error"))
+    | opamfile -> opamfile
+
     let of_channel filename (ic:in_channel) =
-      Pp.parse X.pp ~pos:(pos_file filename) (Syntax.of_channel filename ic)
+      let opamfile = Syntax.of_channel filename ic |> catch_future_syntax_error in
+      Pp.parse X.pp ~pos:(pos_file filename) opamfile
       |> snd
 
     let to_channel filename oc t =
       Syntax.to_channel filename oc (to_opamfile filename t)
 
     let of_string (filename:filename) str =
-      Pp.parse X.pp ~pos:(pos_file filename) (Syntax.of_string filename str)
+      let opamfile = Syntax.of_string filename str |> catch_future_syntax_error in
+      Pp.parse X.pp ~pos:(pos_file filename) opamfile
       |> snd
 
     let to_string filename t =

--- a/src_ext/Makefile.sources
+++ b/src_ext/Makefile.sources
@@ -48,8 +48,8 @@ MD5_0install-solver = 50daf035b04b29399a3c6e6f965ac447
 
 $(call PKG_SAME,0install-solver)
 
-URL_opam-file-format = https://github.com/ocaml/opam-file-format/archive/2.1.2.tar.gz
-MD5_opam-file-format = f16ed774167fc1881876b8185087e0ab
+URL_opam-file-format = https://github.com/ocaml/opam-file-format/archive/2.1.3.tar.gz
+MD5_opam-file-format = b805562dd2d86fc3c8e6d47884fd1da6
 
 $(call PKG_SAME,opam-file-format)
 

